### PR TITLE
GO bot refactored

### PR DESCRIPTION
**Pull request by oserikov**
this PR contains the global GO-bot logic refactoring.
* **removed `obs_size` config param**: the model' input size is now derived from the params.
* NLU logic moved to NLUManager
* text vectorization moved to TextVectorizer
* NLG logic moved to NLGManager
* policy NN logic moved to PolicyNetwork class
* DTO classes introduced for training data entries, configs and shared params

**requires more testing before merge**

---

found bugs (not fixed yet):
* possible GPU bug when gobot_best config is used
* possible bug when GLOVE embeddings are used

todo:
* logic
  * backward compatibility of configs and model serialization\deserialization
  * test on every config with and without gpu, log both execution time and performance
  * simplify config logic: remove conflicting params
  * configs could be stored in properly serializable objects but not dicts
  * allow to easily configure logging level for parts of GO-bot pipeline
* architecture
  * proper OOP for trackers and multiple users tracking
  * remove redundant code duplication in features and targets classes
  * meaningful typehinting for NLU and NLG: trackers, intent extractor, slotfiller, embedders are more than just `Component`s
* improvements
  * simplify templates files composition
  * remove redundant code and datasets duplication for dstc2 cases

### CODE REVIEW

File: deeppavlov/configs/go_bot/gobot_simple_dstc2.json [hunk 1]
```diff
@@ -119,8 +125,8 @@
         "subdir": "{DOWNLOADS_PATH}/embeddings"
       },
       {
-        "url": "http://files.deeppavlov.ai/datasets/simple_dstc2.tar.gz",
```
Comment (oserikov) [inline deeppavlov/configs/go_bot/gobot_simple_dstc2.json:122]: это я наверное погорячился и стоит оставить simple

File: deeppavlov/configs/go_bot/gobot_dstc2.json [hunk 2]
```diff
@@ -52,6 +52,12 @@
           "primary_keys": ["name"],
           "save_path": "{DOWNLOADS_PATH}/dstc2/resto.sqlite"
         },
+        "nlg_manager": {
```
Comment (oserikov) [inline deeppavlov/configs/go_bot/gobot_dstc2.json:55]: all the gobot configs are updated like

before:
* nlg settings are mixed with the rest of gobot settings
now:
* nlg config is now the gobot config field storing all the NLG-specific settings together.

e.g.:
before
```
{
    ...,
    "template_path": ..., <-- NLG
    "slot_filler": ....   <-- not NLG
    ...
    "template_type": ..., <-- not NLG
}
```
after
```
{
    ...,
    "nlg_manager": {
        "class_name": "gobot_nlg_manager",
        "template_path": "{DATA_PATH}/dstc2-templates.txt",
        "template_type": "DualTemplate",
        "api_call_action": "api_call"
    },
    ...,
    "slot_filler": ....   <-- not NLG
}
```

File: deeppavlov/core/common/metrics_registry.json [hunk 3]
```diff
@@ -16,6 +16,7 @@
   "ner_f1": "deeppavlov.metrics.fmeasure:ner_f1",
   "ner_token_f1": "deeppavlov.metrics.fmeasure:ner_token_f1",
   "per_item_bleu": "deeppavlov.metrics.bleu:per_item_bleu",
+  "per_item_action_accuracy": "deeppavlov.metrics.accuracy:per_item_action_accuracy",
```
Comment (oserikov) [inline deeppavlov/core/common/metrics_registry.json:19]: to measure the json-nlg quality

File: deeppavlov/models/go_bot/nlg/mock_json_nlg_manager.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+from typing import Union, Dict
+
+from deeppavlov.core.commands.utils import expand_path
+from deeppavlov.core.common.registry import register
+from deeppavlov.dataset_readers.dstc2_reader import DSTC2DatasetReader
+from deeppavlov.models.go_bot.nlg.dto.json_nlg_response import JSONNLGResponse
+from deeppavlov.models.go_bot.nlg.nlg_manager import log
+from deeppavlov.models.go_bot.nlg.nlg_manager_interface import NLGManagerInterface
+
+
```
Comment (oserikov) [inline deeppavlov/models/go_bot/nlg/mock_json_nlg_manager.py:13]: `decode_response` is the purpose of the class.

File: deeppavlov/models/go_bot/nlg/nlg_manager_interface.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+from abc import ABCMeta, abstractmethod
+
+from deeppavlov.models.go_bot.nlg.dto.nlg_response_interface import NLGResponseInterface
+
+
+class NLGManagerInterface(metaclass=ABCMeta):
```
Comment (oserikov) [inline deeppavlov/models/go_bot/nlg/nlg_manager_interface.py:6]: the idea is that any nlg component should implement this interface

File: deeppavlov/models/go_bot/nlg/nlg_manager.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
+import re
+from logging import getLogger
+from pathlib import Path
+from typing import Union
+
+from deeppavlov.core.commands.utils import expand_path
+import deeppavlov.models.go_bot.nlg.templates.templates as go_bot_templates
```
Comment (yoptar) [inline deeppavlov/models/go_bot/nlg/nlg_manager.py:7]: Optimize imports where possible

File: deeppavlov/configs/go_bot/dstc2/actions2slots.json [hunk 7]
```diff
@@ -0,0 +1,126 @@
+{
```
Comment (yoptar) [inline deeppavlov/configs/go_bot/dstc2/actions2slots.json:1]: This file should not be in configs

File: deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
+# Copyright 2017 Neural Networks and Deep Learning lab, MIPT
```
[the remark this answers is not in the document]
Comment (oserikov) [reply deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:1]: это переименованный файл в котором уже был копирайт, так что оставлю.

[the remark this answers is not in the document]
Comment (yoptar) [reply deeppavlov/models/go_bot/tracker/dialogue_state_tracker.py:1]: Везде нужны копирайты

**Comment by oserikov** [comment]
@dilyararimovna dto ~ data transfer object. I use these to transfer the data between the components, it helps to static code analysis tools and is a bit more powerful than the named tuples.

**Pull request merged**